### ISSUE TITLE
vhost dev has no tx queue, enable failed

### ISSUE THREAD

**Issue by arafatms** [cloud-hypervisor/cloud-hypervisor#2911]
**Describe the bug**
im trying to test vhost-user-net in cloud-hypervisor with DPVGW as the backend. the following error is reported by dpvgw during the negotiation:
vhost dev has no tx queue, enable failed

also got the wrong queue map:
# vgwadm dpdk queue-map
+----------------------------------------------------------------------+-----+----------------------+
|                                ifname                                | tid |      queue_map              |
+----------------------------------------------------------------------+-----+----------------------+
| test_user_sn1.sock                                          |  7  |         0 1 0                        |
| test_user_sn1.sock                                          |  8  |          0 3         |
+----------------------------------------------------------------------+-----+----------------------+ 

As a comparison, the queue created by qemu is like this：
| test_user_sn3.sock |  7  |   0 1 4 5 8 9 12 13  |
| test_user_sn3.sock |  8  |  2 3 6 7 10 11 14 15 |

I also found an interesting thing, qemu will create (queue_num * 2) + cvq virtqueue, while cloudHypervisor will only create queue_num + cvq. Is this problem caused by feature differences?

**To Reproduce**
Steps to reproduce the behaviour:
1.start cloud hypervisor daemon
2.create vm:
```
curl --unix-socket /tmp/test-clh -i \
     -X PUT 'http://localhost/api/v1/vm.create'  \
     -H 'Accept: application/json'               \
     -H 'C
     ontent-Type: application/json'         \
     -d '{
         "cpus":{"boot_vcpus": 8, "max_vcpus": 16},
         "memory":{"size": 4294967296,"hugepages":true,"shared":true,"hotplug_size":4294967296},
         "kernel":{"path":"/vm_data/cloud-hypervisor/vmlinux"},
         "cmdline":{"args":"keep_bootcon console=ttyS0 reboot=k panic=1 root=/dev/pmem0p1 rw iommu=off"},
         "pmem":[{"file":"/vm_data/cloud-hypervisor/centos_7.6.raw","discard_writes":true}],
         "console":{"mode":"Pty"},
         "net":[{"tap":"ich0"}],
         "serial":{"mode":"File","file":"/tmp/clh-serial-test"}
         }'
```

3.boot and attach vhost-user-net device
```
curl --unix-socket /tmp/test-clh -i \
     -X PUT 'http://localhost/api/v1/vm.add-net'  \
     -H 'Accept: application/json'               \
     -H 'C
     ontent-Type: application/json'         \
     -d '{
         "mac":"52:54:00:dc:19:f8","vhost_user":true,"vhost_mode":"Server","vhost_socket":"test_user_sn1.sock","num_queues":8,"queue_size":128
         }'
```

**Version**

Output of `cloud-hypervisor --version`:
cloud-hypervisor v17.0.0

Did you build from source, if so build command line (e.g. features):
rpm build -ba cloud-hypervisor.spec

**VM configuration**

What command line did you run (or JSON config data):

Guest OS version details:
Kernel 5.6.0-rc4+ on an x86_64

Host OS version details:
4.18.0-193.1.0.1

**Logs**

Output of `cloud-hypervisor -v` from either standard error or via `--log-file`:
without any errors
```
cloud-hypervisor: 2289.342939924s: <vmm> INFO:vmm/src/lib.rs:1161 -- API request event: VmAddNet(NetConfig { tap: None, ip: 192.168.249.1, mask: 255.255.255.0, mac: MacAddr { bytes: [82, 84, 0, 220, 25, 248] }, host_mac: None, iommu: false, num_queues: 8, queue_size: 128, vhost_user: true, vhost_socket: Some("test_user_sn1.sock"), vhost_mode: Server, id: None, fds: None, rate_limiter_config: None }, Sender { .. })
cloud-hypervisor: 2289.343067209s: <vmm> INFO:vmm/src/device_manager.rs:1978 -- Creating virtio-net device: NetConfig { tap: None, ip: 192.168.249.1, mask: 255.255.255.0, mac: MacAddr { bytes: [82, 84, 0, 220, 25, 248] }, host_mac: None, iommu: false, num_queues: 8, queue_size: 128, vhost_user: true, vhost_socket: Some("test_user_sn1.sock"), vhost_mode: Server, id: Some("_net4"), fds: None, rate_limiter_config: None }
cloud-hypervisor: 2289.343165393s: <vmm> INFO:virtio-devices/src/vhost_user/vu_common_ctrl.rs:269 -- Binding vhost-user listener...
cloud-hypervisor: 2289.343210387s: <vmm> INFO:virtio-devices/src/vhost_user/vu_common_ctrl.rs:271 -- Waiting for incoming vhost-user connection...
cloud-hypervisor: 2289.579372928s: <_net0_qp0> DEBUG:vm-virtio/src/queue.rs:701 -- used_event = 1170 used_idx = 1171 old_idx = 1170
cloud-hypervisor: 2289.579407686s: <_net0_qp0> DEBUG:vm-virtio/src/queue.rs:712 -- Needs notification: true
cloud-hypervisor: 2289.579430294s: <_net0_qp0> DEBUG:virtio-devices/src/net.rs:212 -- Signalling RX queue
cloud-hypervisor: 2290.102899093s: <vmm> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
cloud-hypervisor: 2290.102982562s: <vcpu0> DEBUG:vmm/src/device_manager.rs:4049 -- PCI_HP_REG_R: base 0x3ffffffee000, offset 0x0, data [32, 0, 0, 0]
cloud-hypervisor: 2290.103105557s: <vcpu0> DEBUG:vmm/src/device_manager.rs:4049 -- PCI_HP_REG_R: base 0x3ffffffee000, offset 0x4, data [0, 0, 0, 0]
cloud-hypervisor: 2290.10595184s: <vcpu2> DEBUG:pci/src/configuration.rs:823 -- DETECT BAR REPROG: current 0x3ffa, new 0x2
cloud-hypervisor: 2290.107117218s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:146 -- write_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107140018s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107160231s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107179356s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:146 -- write_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107219012s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107238924s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:146 -- write_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107257938s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:221 -- write_common_config_dword: offset 0x0
cloud-hypervisor: 2290.107277177s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:193 -- read_common_config_dword: offset 0x4
cloud-hypervisor: 2290.107296984s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:221 -- write_common_config_dword: offset 0x0
cloud-hypervisor: 2290.107316718s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:193 -- read_common_config_dword: offset 0x4
cloud-hypervisor: 2290.107344477s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:221 -- write_common_config_dword: offset 0x8
cloud-hypervisor: 2290.107364016s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:221 -- write_common_config_dword: offset 0xc
cloud-hypervisor: 2290.107383517s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:221 -- write_common_config_dword: offset 0x8
cloud-hypervisor: 2290.107410868s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:221 -- write_common_config_dword: offset 0xc
cloud-hypervisor: 2290.107429689s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107448814s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:146 -- write_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107467521s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x14
cloud-hypervisor: 2290.107497981s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x15
cloud-hypervisor: 2290.10753839s: <vcpu2> DEBUG:virtio-devices/src/transport/pci_common_config.rs:133 -- read_common_config_byte: offset 0x15
cloud-hypervisor: 2290.107650358s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x0 data 0xfee00000
cloud-hypervisor: 2290.107672564s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x4 data 0x0
cloud-hypervisor: 2290.107694505s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x8 data 0x40ef
cloud-hypervisor: 2290.107719207s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x10 data 0xfee00000
cloud-hypervisor: 2290.107740375s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x14 data 0x0
cloud-hypervisor: 2290.107761011s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x18 data 0x40ef
cloud-hypervisor: 2290.107782933s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x20 data 0xfee00000
cloud-hypervisor: 2290.107803422s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x24 data 0x0
cloud-hypervisor: 2290.107823921s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x28 data 0x40ef
cloud-hypervisor: 2290.107845546s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x30 data 0xfee00000
cloud-hypervisor: 2290.107866138s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x34 data 0x0
cloud-hypervisor: 2290.107886691s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x38 data 0x40ef
cloud-hypervisor: 2290.107908385s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x40 data 0xfee00000
cloud-hypervisor: 2290.107928813s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x44 data 0x0
cloud-hypervisor: 2290.107950377s: <vcpu2> DEBUG:pci/src/msix.rs:279 -- MSI_W TABLE offset 0x48 data 0x40ef
```

vow log:
```
vgw: VHOST_CONFIG: negotiated Virtio features: 0x16042ff83
vgw: VHOST_CONFIG: read message VHOST_USER_SET_MEM_TABLE
vgw: VHOST_CONFIG: guest memory region 0, size: 0x20000000#012#011 guest physical addr: 0x0#012#011 guest virtual  addr: 0x7fffd7800000#012#011 host  virtual  addr: 0x7fffa0000000#012#011 mmap addr : 0x7fffa0000000#012#011 mmap size : 0x20000000#012#011 mmap align: 0x200000#012#011 mmap off  : 0x0
vgw: VHOST_CONFIG: guest memory region 1, size: 0x20000000#012#011 guest physical addr: 0x100000000#012#011 guest virtual  addr: 0x7ffab5a00000#012#011 host  virtual  addr: 0x7fdd20000000#012#011 mmap addr : 0x7fdd20000000#012#011 mmap size : 0x20000000#012#011 mmap align: 0x200000#012#011 mmap off  : 0x0
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_NUM
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_ADDR
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_BASE
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_CALL
vgw: VHOST_CONFIG: vring call idx:0 file:204
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_KICK
vgw: VHOST_CONFIG: vring kick idx:0 file:205
vgw: VHOST_CONFIG: reallocate vq from 0 to 1 node
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_ENABLE
vgw: VHOST_CONFIG: set queue enable: 1 to qp idx: 0
vgw: WARNING__vhost_user_vring_enable__235[@thread-0]: vring(0) enable(1) failed, can not find nval_device(test_user_sn1.sock)!
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_NUM
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_ADDR
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_BASE
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_CALL
vgw: VHOST_CONFIG: vring call idx:1 file:206
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_KICK
vgw: VHOST_CONFIG: vring kick idx:1 file:207
vgw: VHOST_CONFIG: reallocate vq from 0 to 1 node
vgw: VHOST_CONFIG: read message VHOST_USER_SET_VRING_ENABLE
vgw: VHOST_CONFIG: set queue enable: 1 to qp idx: 1
vgw: WARNING__vhost_user_vring_enable__235[@thread-0]: vring(1) enable(1) failed, can not find nval_device(test_user_sn1.sock)!
vgw: VHOST_CONFIG: virtio is now ready for processing.
vgw: LOG__vhost_user_new_device__380[@thread-0]: vhost user new_device ifname test_user_sn1.sock
vgw: LOG__vhost_user_new_device__411[@thread-0]: vhost user new_device init tx_queue 0 succeed
vgw: LOG__vhost_user_vring_enable__239[@thread-0]: vring_enable queue 0 enabled 1(vid:2 socket_path:test_user_sn1.sock)
vgw: LOG__vhost_user_vring_enable__239[@thread-0]: vring_enable queue 1 enabled 1(vid:2 socket_path:test_user_sn1.sock)
VHOST_CONFIG: read message VHOST_USER_SET_VRING_NUM
VHOST_CONFIG: read message VHOST_USER_SET_VRING_ADDR
VHOST_CONFIG: read message VHOST_USER_SET_VRING_BASE
VHOST_CONFIG: read message VHOST_USER_SET_VRING_CALL
VHOST_CONFIG: vring call idx:2 file:208
VHOST_CONFIG: read message VHOST_USER_SET_VRING_KICK
VHOST_CONFIG: vring kick idx:2 file:209
VHOST_CONFIG: read message VHOST_USER_SET_VRING_ENABLE
vgw: VHOST_CONFIG: set queue enable: 1 to qp idx: 2
vgw: LOG__vhost_user_vring_enable__239[@thread-0]: vring_enable queue 2 enabled 1(vid:2 socket_path:test_user_sn1.sock)
vgw: ERROR__vhost_user_vring_enable__254[@thread-0]: vhost dev has no tx queue 2, enable failed
```

**Comment by rbradford** [comment]
In Cloud Hypervisor `num_queues` is not the number of queue pairs but rather the number of queue_pairs * 2.

**Comment by sboeuf** [comment]
@arafatms 
So I think your HTTP request should be:
```
"mac":"52:54:00:dc:19:f8","vhost_user":true,"vhost_mode":"Server","vhost_socket":"test_user_sn1.sock","num_queues":16,"queue_size":128
```

**Comment by likebreath** [comment]
Fixed by #2921.
